src/test/neorados/common_tests.h

```diff
@@ -15,27 +15,141 @@
+	    }
+	    co_return;
+	  }, ba::detached);
+      }, token, executor, r.get_executor());
+}
+
+template<typename CompletionToken>
+auto create_obj(neorados::RADOS& r, std::string_view oid,
+		const neorados::IOContext& ioc,
+		CompletionToken&& token)
+{
+  neorados::WriteOp op;
+  op.create(true);
+  return r.execute(oid, ioc, std::move(op), std::forward<CompletionToken>(token));
+}
+
+class temp_pool {
+  neorados::RADOS* r = nullptr;
+  neorados::IOContext c;
+  std::optional<
+    boost::asio::executor_work_guard<neorados::RADOS::executor_type>> w;
+
+public:
+
+  temp_pool() = default;
+  temp_pool(neorados::RADOS& r_, const int64_t pid)
+    : r(&r_), c(neorados::IOContext(pid)),
+      w(boost::asio::make_work_guard(r->get_executor())) {
+    std::cerr << "Construct temp_pool " << std::hex << uintptr_t(this) << std::dec << std::endl;
+  }
+
+  temp_pool(const temp_pool&) = delete;
+  temp_pool& operator =(const temp_pool&) = delete;
+
+  temp_pool(temp_pool&& rhs)
+    : r(rhs.r), c(std::move(rhs.c)) {
+    std::cerr << "Moving temp_pool from " << std::hex << uintptr_t(&rhs)
+	      << " to " << uintptr_t(this) << std::dec << std::endl;
+    rhs.r = nullptr;
+    rhs.w.reset();
+  }
+  temp_pool& operator =(temp_pool&& rhs) {
+    std::cerr << "Move assigning temp_pool from " << std::hex << uintptr_t(&rhs)
+	      << " to " << uintptr_t(this) << std::dec << std::endl;
+    r = rhs.r;
+    rhs.r = nullptr;
+    c = std::move(rhs.c);
+    w.reset();
+    w.emplace(boost::asio::make_work_guard(r->get_executor()));
+    rhs.w.reset();
+    return *this;
+  }
+
+  ~temp_pool() {
+    std::cerr << "Destructing temp_pool " << std::hex << uintptr_t(this) << std::endl;
+    if (r) {
+      std::cout << "Firing delete for pool " << c.pool() << std::endl;
+      r->delete_pool(c.pool(), boost::asio::detached);
+    }
+    std::cerr.flush();
+  }
+
+  operator const neorados::IOContext& () {
+    return c;
+  }
+};
+
+template<typename CompletionToken>
+auto make_temp_pool(
+  neorados::RADOS& r,
+  CompletionToken&& token)
 {
-  boost::asio::async_completion<CompletionToken,
-				void(boost::system::error_code,
-				     std::int64_t)> init(token);
-  r.create_pool(pname, std::nullopt,
-		[&r, pname = std::string(pname),
-		 h = std::move(init.completion_handler)]
-		(boost::system::error_code ec) mutable {
-		  r.lookup_pool(
-		    pname,
-		    [h = std::move(h)]
-		    (boost::system::error_code ec, std::int64_t pool) mutable {
-		      std::move(h)(ec, pool);
-		    });
-		});
-  return init.result.get();
+  namespace ba = boost::asio;
+  namespace bs = boost::system;
+  auto&& executor = ba::get_associated_executor(token);
+  std::cerr << "Calling make_temp_pool." << std::endl;
+  return ba::async_compose<CompletionToken, void(bs::error_code, temp_pool)>
+    (
+      [&executor, &r](auto& self) {
+	std::cerr << "In async_compose" << std::endl;
+	co_spawn(
+	  executor,
+	  [&r, &self]() mutable -> ba::awaitable<void> {
```

Comment:
ok, the problem is the lifetime of `self` here. it goes out of scope and destructs right after calling `co_spawn()`

you need to preserve its lifetime by moving `self` into the coroutine's lambda capture